lgc/elfLinker/ElfLinker.cpp

```diff
@@ -409,6 +417,37 @@ bool ElfLinkerImpl::link(raw_pwrite_stream &outStream) {
     }
   }
 
+  // Add relocations that cannot be applied at this stage.
+  for (auto &elfInput : m_elfInputs) {
+    for (const object::SectionRef section : elfInput.objectFile->sections()) {
+      unsigned sectType = object::ELFSectionRef(section).getType();
+      if (sectType == ELF::SHT_REL || sectType == ELF::SHT_RELA) {
+        for (object::RelocationRef reloc : section.relocations()) {
+          unsigned targetSectionIdx = UINT_MAX;
+          unsigned targetIdxInSection = UINT_MAX;
+          std::tie(targetSectionIdx, targetIdxInSection) =
+              findInputSection(elfInput, *cantFail(section.getRelocatedSection()));
+          if (targetSectionIdx != UINT_MAX) {
+            uint64_t value = 0;
+            if (getRelocValue(reloc, value)) {
+              continue;
+            }
+            assert(targetSectionIdx == textSectionIdx && "We assume all relocations are applied to the text section");
+            assert(sectType == ELF::SHT_REL && "We do not output a RELA section yet");
+            object::SectionRef relocSection = *cantFail(reloc.getSymbol()->getSection());
+            unsigned relocSectionId = UINT_MAX;
+            unsigned relocIdxInSection = UINT_MAX;
+            std::tie(relocSectionId, relocIdxInSection) = findInputSection(elfInput, relocSection);
+            uint64_t relocSectionOffset = m_outputSections[relocSectionId].getOutputOffset(relocIdxInSection);
+            uint64_t targetSectionOffset = m_outputSections[targetSectionIdx].getOutputOffset(targetIdxInSection);
+            m_outputSections[relocSectionId].addRelocation(reloc, elfInput.objectFile->getFileName(),
+                                                           relocSectionOffset, targetSectionOffset);
+          }
+        }
+      }
+    }
+  }
```

Comment:
Would it be possible to merge this loop with the loop that applies relocations? It seems strange to loop over all input relocations twice.

Reply:
It would be possible but then some else ackward will have to happen.  We need to output the text section before we can apply the relocations.  However, we want to get the relocations to output before we output all of the sections.  If we wanted to skip the relocations when we output all of the other section and output it later, like we with with the .note section, then we could merge the loops.

Reply:
Wouldn't a linker typically output the relocations at the end? So it would make sense to postpone the writing of relocation sections similar to the .note section? Maybe there's something else I'm not seeing yet.

Reply:
I just tried doing it later again, and there is something else.  We create new symbols that point to the data in the .rodata section when we add the new relocations.  This way we only create a new symbol if it is needed.  So we have to output the symbol table and string table after we have added the relocations.  If we were to do only loop for the relocations the order would be

So the order would have to be:

1. output the text section
1. process and apply relocations
1. output the sym table, string table, relocations, and note sections.

We could change things so that we output just the text section first, and then we do the general loop to output all of the section after the relocations have been processed.

@trenouf Do you have a opinion on the order in which things are done?  This is your code.

Reply:
My preference is to leave it as you have it here with a separate loop. There aren't that many relocs.